### PR TITLE
Port `g` (egg/shellcode) commands to the rzshell

### PR DESCRIPTION
**Pull request by XVilka**
**Your checklist for this pull request**
- [x] I've read the [guidelines for contributing](https://github.com/rizinorg/rizin/blob/master/DEVELOPERS.md) to this repository
- [x] I made sure to follow the project's [coding style](https://github.com/rizinorg/rizin/blob/master/DEVELOPERS.md#code-style)
- [ ] I've added tests that prove my fix is effective or that my feature works (if possible)
- [ ] I've updated the documentation and the [rizin book](https://github.com/rizinorg/book) with the relevant information (if needed)

**Detailed description**

- Migrate `g` (egg/shellcode) commands to the rzshell
- Removed `gw`, since it was mentioned in the help but not implemented
- Made `gc` (set egg options) behave like `e` command

**Test plan**

CI is green

Partially addresses #1342

### CODE REVIEW

File: librz/core/cmd_descs/cmd_egg.yaml [hunk 1]
```diff
@@ -0,0 +1,71 @@
...
+    args: []
+  - name: gs
+    cname: egg_syscall
+    summary: Compile syscall "name(args)"
+    args:
+      - name: name
+        type: RZ_CMD_ARG_TYPE_STRING
+      - name: args
+        type: RZ_CMD_ARG_TYPE_STRING
+        optional: true
```
Comment (ret2libc) [inline librz/core/cmd_descs/cmd_egg.yaml:43]: shall this be an ARRAY?

**Review by ret2libc — CHANGES_REQUESTED** [review]
This needs at least few tests before refactoring it. 
See https://codecov.io/gh/rizinorg/rizin/src/a527b204d8573fea3d3aabc6d791334202928f85/librz/core/cmd_egg.c . It's almost entirely untested, so we can't reliably change the code.

File: librz/core/cmd_descs/cmd_egg.yaml (hunk 1 above)
Comment (XVilka) [reply librz/core/cmd_descs/cmd_egg.yaml:43]: There is no `RZ_CMD_ARG_TYPE_ARRAY` exists.

Comment (ret2libc) [reply librz/core/cmd_descs/cmd_egg.yaml:43]: see `flags` in cmd_descs.yaml . RZ_CMD_ARG_FLAG_ARRAY

Comment (XVilka) [reply librz/core/cmd_descs/cmd_egg.yaml:43]: Well, it's not flags, it's syscall arguments. Not sure if it's appropriate here.

Comment (ret2libc) [reply librz/core/cmd_descs/cmd_egg.yaml:43]: yes, but it's more than one. If you set `flags: RZ_CMD_ARG_FLAG_ARRAY` for the argument, RzCmd will know that it's an array of elements. Grep for it in the code, you'll see.

Comment (XVilka) [reply librz/core/cmd_descs/cmd_egg.yaml:43]: Done.

File: librz/core/cmd_egg.c [hunk 2]
```diff
@@ -131,166 +90,182 @@ static int cmd_egg_compile(RzEgg *egg) {
...
+		return NULL;
+	}
+	return egg;
+}
+
+RZ_IPI RzCmdStatus rz_egg_compile_handler(RzCore *core, int argc, const char **argv) {
+	RzEgg *egg = rz_core_egg_setup(core);
+	if (!egg) {
+		return RZ_CMD_STATUS_ERROR;
+	}
```
Comment (ret2libc) [inline librz/core/cmd_egg.c]: Is this really needed every time you execute any `g` command?

File: librz/core/cmd_egg.c [hunk 3]
```diff
@@ -131,166 +90,182 @@ static int cmd_egg_compile(RzEgg *egg) {
...
+}
+
+RZ_IPI RzCmdStatus rz_egg_config_handler(RzCore *core, int argc, const char **argv) {
+	RzEgg *egg = rz_core_egg_setup(core);
+	if (!egg) {
+		return RZ_CMD_STATUS_ERROR;
+	}
+	if (argc > 1) {
+		// Set the config option
+		RzList *l = rz_str_split_duplist_n(argv[1], "=", 1, false);
```
Comment (ret2libc) [inline librz/core/cmd_egg.c]: I think this is wrong. You potentially have an array of arguments, not just `argv[1]`, so each one of them should be split by `=`, not just `argv[1]`.

**Review by ret2libc — CHANGES_REQUESTED** [review]
Before doing any other kind of changes please add even very very basic tests under `db` for rz-gg or `g` commands, otherwise we'll never know if this work. As said, it's not tested at all.

File: librz/core/cmd_egg.c (hunk 2 above)
Comment (XVilka) [reply librz/core/cmd_egg.c]: It was how old logic worked.

Comment (ret2libc) [reply librz/core/cmd_egg.c]: kk

**Comment by XVilka** [comment]
Added tests and unified some code with `rz-gg` tool, please take a look again.

**Comment by XVilka** [comment]
The failing test seems also to be failing for the `rz-gg` tool itself, which shares the same code with `g` commands. Test expects the shellcode with the last `jmp 0x60` instruction:
```
[0x00000000]> pd
            0x00000000      55             push  ebp
            0x00000001      89e5           mov   ebp, esp
            0x00000003      81ec80000000   sub   esp, 0x80
            0x00000009      c74508686921.  mov   dword [ebp + 8], 0xa216968
            0x00000010      c7450c6e0000.  mov   dword [ebp + 0xc], 0x6e ; 'n'
            0x00000017      c74510000000.  mov   dword [ebp + 0x10], 0
            0x0000001e      8d4508         lea   eax, [ebp + 8]
            0x00000021      898504000000   mov   dword [ebp + 4], eax
            0x00000027      c74505010000.  mov   dword [ebp + 5], 1
            0x0000002e      c74506020000.  mov   dword [ebp + 6], 2
            0x00000035      55             push  ebp
            0x00000036      58             pop   eax
            0x00000037      3b4506         cmp   eax, dword [ebp + 6]
        ╭─< 0x0000003a      0f8f5a000000   jg    0x9a
        │   0x00000040      6a04           push  4
        │   0x00000042      54             push  esp
        │   0x00000043      6a01           push  1
        │   0x00000045      8b1c24         mov   ebx, dword [esp]
        │   0x00000048      8b8c24040000.  mov   ecx, dword [esp + 4]
        │   0x0000004f      8b9424080000.  mov   edx, dword [esp + 8]
        │   0x00000056      b804000000     mov   eax, 4
        │   0x0000005b      cd80           int   0x80
        │   0x0000005d      83c40c         add   esp, 0xc
       ╭──> 0x00000060      6a00           push  0
       ╎│   0x00000062      8b1c24         mov   ebx, dword [esp]
       ╎│   0x00000065      b801000000     mov   eax, 1
       ╎│   0x0000006a      cd80           int   0x80
       ╎│   0x0000006c      83c404         add   esp, 4
       ╎│   0x0000006f      81c480000000   add   esp, 0x80
       ╎│   0x00000075      5d             pop   ebp
       ╎│   0x00000076      c3             ret
       ╰──< 0x00000077      ebe7           jmp   0x60
```
Here is the source code of this `simple_cmp.r` shellcode:
```
write@syscall(4);
exit@syscall(1);

main@global(128) {
  .var0 = "hi!\n";
  .var1 = 1;
  .var2 = 2;
  if (.var1 < .var2) {
    write(1,.var0, 4);
  }
  exit(0);
}
```
The assembly code that `simple_cmp.r` produces is this:
```
.global main
main:
  push ebp
  mov ebp, esp
  sub esp, 128
# encode filtered string (hi!\n) ((null))
  mov dword [ebp+8], 0xa216968
  mov dword [ebp+12], 0x0
  mov dword [ebp+16], 0
  lea eax, [ebp+8]
  mov [ebp+4], eax
  mov ebx, 0
  mov ebx, 1
  mov [ebp+5], ebx
  mov ebx, 0
  mov ebx, 2
  mov [ebp+6], ebx
# rcc_fun 1 (if)
  push [ebp+4]
  __begin_0_1_0:
# cond frame if (var1 < .var2)
  pop eax
  cmp eax, [ebp+6]
  jg __end_0_1_0
# rcc_fun 2 (write)
  push 4
  push [ebp+4]
  push 1
# set syscall args
  mov ebx, [esp]
  mov ecx, [esp+4]
  mov edx, [esp+8]
# syscall
 mov eax, 4
 int 0x80
  add esp, 12
  jmp __ifelse_1_0
  __end_0_1_0:
# rcc_fun 1 (exit)
  push 0
# set syscall args
  mov ebx, [esp]
# syscall
 mov eax, 1
 int 0x80
  add esp, 4
  add esp, 128
  pop ebp
  ret

  __ifelse_1_0:
  jmp __end_0_1_0
```

File: librz/core/cmd/cmd_egg.c [hunk 4]
```diff
@@ -131,166 +94,200 @@ static int cmd_egg_compile(RzEgg *egg) {
...
+
+RZ_IPI RzCmdStatus rz_egg_syscall_handler(RzCore *core, int argc, const char **argv) {
+	RzEgg *egg = rz_core_egg_setup(core);
+	if (!egg) {
+		return RZ_CMD_STATUS_ERROR;
+	}
+	RzBuffer *buf = NULL;
+	if (argc > 2) {
+		// With syscall parameters specified
+		buf = rz_core_syscall(core, argv[1], argv[2]);
```
Comment (ret2libc) [inline librz/core/cmd/cmd_egg.c:222]: `argv[2]` is marked as FLAG_ARRAY but then you just use `argv[2]`. I think you should not use any flag for that argument. That way the parser will automatically compact everything in one single string.

**Comment by ret2libc** [comment]
Just set back that test as BROKEN=1. It seems it was already broken before your PR.

**Comment by XVilka** [comment]
Yes, I was just wanting to test the `g` command with `*.r` file. But nevetheless, the result is the same as with `rz-gg` command. I will add the same tests from `rz-gg` that aren't broken

**Review by ret2libc — APPROVED** [review]

**Pull request merged**